Client/Frontend/Browser/DownloadQueue.swift

```diff
@@ -0,0 +1,175 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+private let downloadOperationQueue = OperationQueue()
+
+protocol DownloadDelegate {
+    func download(_ download: Download, didCompleteWithError error: Error?)
+    func download(_ download: Download, didDownloadBytes bytesDownloaded: Int64)
+    func download(_ download: Download, didFinishDownloadingTo location: URL)
+}
+
+class Download: NSObject {
+    let preflightResponse: URLResponse
+    let request: URLRequest
+
+    let mimeType: String
+    let filename: String
+
+    var delegate: DownloadDelegate?
+
+    var state: URLSessionTask.State {
+        return task?.state ?? .suspended
+    }
+
+    private(set) var totalBytesExpected: Int64?
+    private(set) var bytesDownloaded: Int64
+
+    private(set) var session: URLSession?
+    private(set) var task: URLSessionDownloadTask?
+
+    init(preflightResponse: URLResponse, request: URLRequest) {
+        self.preflightResponse = preflightResponse
+        self.request = request
+
+        self.mimeType = preflightResponse.mimeType ?? "application/octet-stream"
+        self.filename = preflightResponse.suggestedFilename ?? "unknown"
+
+        self.totalBytesExpected = preflightResponse.expectedContentLength
+        self.bytesDownloaded = 0
+
+        super.init()
+
+        self.session = URLSession(configuration: .default, delegate: self, delegateQueue: downloadOperationQueue)
```

Comment:
should we be setting the Firefox UserAgent here?

Reply:
Doesn't seem necessary. Since we're hijacking the `URLRequest` from the `WKNavigationDelegate`, it already has all the necessary headers (including `User-Agent`) set correctly.